crates/sui-types/src/crypto.rs

```diff
@@ -262,77 +266,207 @@ impl std::fmt::Debug for Signature {
+}
+
+impl AsRef<[u8]> for Ed25519SuiSignature {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
     }
+}
 
-    pub fn new_empty() -> Self {
-        Self([0u8; SUI_SIGNATURE_LENGTH])
+impl signature::Signature for Ed25519SuiSignature {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != Self::LENGTH {
+            return Err(signature::Error::new());
+        }
+        let mut sig_bytes = [0; Self::LENGTH];
+        sig_bytes.copy_from_slice(bytes);
+        Ok(Self(sig_bytes))
     }
+}
+
+// impl signature::Signer<Signature> for Ed25519KeyPair {
+//     fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+//         Ok(Ed25519SuiSignature::new(self, msg)
+//             .map_err(|_| signature::Error::new())?.into())
+//     }
+// }
 
-    pub fn signature_bytes(&self) -> &[u8] {
-        &self.0[..AccountSignature::LENGTH]
+//
+// Ed25519 Sui Signature port
+//
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct Secp256k1SuiSignature(
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "Readable<Base64, Bytes>")]
+    [u8; Secp256k1PublicKey::LENGTH + Secp256k1Signature::LENGTH + FLAG_LENGTH],
+);
+
+impl SuiSignatureInner for Secp256k1SuiSignature {
+    type Sig = Secp256k1Signature;
+    type PubKey = Secp256k1PublicKey;
+    type KeyPair = Secp256k1KeyPair;
+    const LENGTH: usize = Secp256k1PublicKey::LENGTH + Secp256k1Signature::LENGTH + FLAG_LENGTH;
+    const FLAG: [u8; FLAG_LENGTH] = [0xed];
+}
+
+impl AsRef<[u8]> for Secp256k1SuiSignature {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
     }
+}
 
-    pub fn public_key_bytes(&self) -> &[u8] {
-        &self.0[AccountSignature::LENGTH..]
+impl signature::Signature for Secp256k1SuiSignature {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != Self::LENGTH {
+            return Err(signature::Error::new());
+        }
+        let mut sig_bytes = [0; Self::LENGTH];
+        sig_bytes.copy_from_slice(bytes);
+        Ok(Self(sig_bytes))
     }
+}
 
-    /// This performs signature verification on the passed-in signature, additionally checking
-    /// that the signature was performed with a PublicKey belonging to an expected author, indicated by its Sui Address
-    pub fn verify<T>(&self, value: &T, author: SuiAddress) -> Result<(), SuiError>
-    where
-        T: Signable<Vec<u8>>,
-    {
-        let mut message: Vec<u8> = Vec::new();
-        value.write(&mut message);
-        let (signature, public_key_bytes) = self.get_verification_inputs(author)?;
+impl signature::Signer<Signature> for Secp256k1KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+        Ok(Secp256k1SuiSignature::new(self, msg)
+            .map_err(|_| signature::Error::new())?
+            .into())
+    }
+}
+
+//
+// This struct exists due to the limitations of the `enum_dispatch` library.
```

Comment:
what is the limitation?

Reply:
We cannot have associated types on structs that are tagged with 'enum_dispatch'